# Hola &#x1F600;,

Soy **Hesus Garcia** – **"Soy el único Hesus que conoces (y probablemente conocerás) 🌟"** – Sí, como "Jesús", pero con una H que me hace único. Puede sonar raro, pero créeme, ¡no lo olvidarás! Como tu revisor en Triple-Ten, estoy aquí para guiarte y ayudarte a mejorar tu código. Si algo necesita un ajuste, no hay de qué preocuparse; ¡aquí estoy para hacer que tu trabajo brille con todo su potencial! ✨

Cada vez que encuentre un detalle importante en tu código, te lo señalaré para que puedas corregirlo y así te prepares para un ambiente de trabajo real, donde el líder de tu equipo actuaría de manera similar. Si en algún momento no logras solucionar el problema, te daré más detalles para ayudarte en nuestra próxima oportunidad de revisión.

Es importante que cuando encuentres un comentario, **no los muevas, no los modifiques, ni los borres**.

---

### Formato de Comentarios

Revisaré cuidadosamente cada implementación en tu notebook para asegurar que cumpla con los requisitos y te daré comentarios de acuerdo al siguiente formato:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> - ¡Excelente trabajo! Esta parte está bien implementada y contribuye significativamente al análisis de datos o al proyecto. Continúa aplicando estas buenas prácticas en futuras secciones.
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Este código está correcto, pero se puede optimizar. Considera implementar mejoras para que sea más eficiente y fácil de leer. Esto fortalecerá la calidad de tu proyecto.
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error en el código que es necesario corregir para aprobar esta sección. Por favor, revisa y corrige este punto, ya que es fundamental para la validez del análisis y la precisión de los resultados.
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor** que incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto.
- **Áreas de mejora:** Sugerencias sobre aspectos donde puedes mejorar.
- **Temas adicionales para investigar:** Ideas de temas opcionales que puedes explorar por tu cuenta para desarrollar aún más tus habilidades.

Estos temas adicionales no son obligatorios en esta etapa, pero pueden serte útiles para profundizar en el futuro.

---


Esta estructura en viñetas facilita la lectura y comprensión de cada parte del comentario final.

También puedes responderme de la siguiente manera si tienes alguna duda o quieres aclarar algo específico:


<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
    
Aquí puedes escribir tu respuesta o pregunta sobre el comentario.
    
</div>


**¡Empecemos!** &#x1F680;


# Proyecto RappiPlus: de datos a decisiones de negocio

**Introducción**


El objetivo de este proyecto es evaluar el desempeño del servicio **RappiPlus** para apoyar **decisiones de negocio basadas en datos**.

Se trabajan con múltiples datasets del negocio:

- **rappiplus_orders_raw.csv** → información de pedidos, precios, descuentos y revenue  
- **rappiplus_catalog.csv** → costos de productos, categorías y proveedores  
- **rappiplus_marketing_spend.csv** → inversión en marketing por canal y país  
- **events / users / user_activity (SQL)** → comportamiento del usuario dentro de la plataforma  
- **experiment_checkout_ui.csv** → resultados de un experimento A/B en el checkout  

El análisis sigue una lógica clara y progresiva:

1. 🔍 Evaluar si podemos confiar en los datos (calidad de datos en Python) 

2. 💰 Analizar si el negocio es rentable (revenue, costos y profit)  

3. 🛒 Entender dónde se pierden los usuarios (funnel de conversión)  

4. 🔁 Evaluar si los usuarios regresan (retención por cohortes)  

5. 🧪 Validar si los cambios generan impacto (test estadístico)  

6. 📊 Comunicar los resultados (dashboard en BI)  

A lo largo del proyecto, se transforman datos en insights para responder preguntas clave del negocio y proponer **recomendaciones accionables**.

---

## 🔹 Paso 1: Cargar y validar la calidad de los datos

---

### 1.1 Carga de datos y vista rápida

**🎯 Objetivo:** Familiarizarte con la estructura de los datasets del negocio antes de analizarlos.

**Instrucciones:**

- Importa las librerías necesarias
- Carga los archivos:
  - `rappiplus_orders_raw.csv`
  - `rappiplus_catalog.csv`
  - `rappiplus_marketing_spend.csv`
- Guarda los DataFrames en:
  - `orders`, `catalog`, `marketing`
- Explora cada dataset.

---

In [197]:
# importar librerías
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [198]:
# cargar archivos
orders = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_orders_raw.csv')
catalog = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_catalog.csv')
marketing = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_marketing_spend.csv')

In [199]:
# explorar datasets
orders.info()
orders.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25100 entries, 0 to 25099
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id_pedido           25100 non-null  object 
 1   id_usuario          25100 non-null  object 
 2   fecha_hora_pedido   25100 non-null  object 
 3   pais                24800 non-null  object 
 4   dispositivo         25080 non-null  object 
 5   fuente_referencia   25070 non-null  object 
 6   nombre_producto     25070 non-null  object 
 7   categoria_producto  25020 non-null  object 
 8   cantidad            25050 non-null  float64
 9   precio_unitario     25050 non-null  float64
 10  monto_descuento     25050 non-null  float64
 11  monto_total         25100 non-null  float64
dtypes: float64(4), object(8)
memory usage: 2.3+ MB


,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
0,order_0,user_6993,2025-05-22,Argentina,desktop,organic,Jacket-Winter-M,Moda,2.0,332.69,0.0,665.37
1,order_1,user_1329,2025-06-15,Mexico,desktop,paid_search,Tablet-Standard-64GB,Electronica,1.0,176.86,5.0,171.86
2,order_2,user_3194,2025-05-02,Argentina,desktop,social,Blender-XL-Red,Hogar,2.0,102.99,10.0,195.99
3,order_3,user_4510,2025-06-09,Colombia,mobile,social,Tablet-Standard-64GB,Electronica,1.0,257.87,15.0,242.87
4,order_4,user_5044,2025-03-30,Argentina,desktop,paid_search,Blender-XL-Red,Hogar,1.0,336.28,0.0,336.28
5,order_5,user_2792,2025-04-22,mexico,desktop,organic,Tablet-Standard-64GB,Electronica,1.0,179.34,0.0,179.34
6,order_6,user_2802,2025-03-15,Colombia,mobile,organic,Blender-XL-Red,Hogar,2.0,163.59,0.0,327.19
7,order_7,user_1083,2025-05-01,Mexico,desktop,organic,Tablet-Standard-64GB,Electronica,2.0,373.68,0.0,747.35
8,order_8,user_2352,2025-03-01,Argentina,desktop,paid_search,Laptop-Gaming-16GB,Electronica,2.0,477.27,5.0,949.54
9,order_9,user_7001,2025-05-31,Mexico,mobile,organic,Sneakers-Urban-42,Moda,1.0,339.30,5.0,334.30


In [200]:
catalog.info()
catalog.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   nombre_producto     7 non-null      object 
 1   categoria_producto  7 non-null      object 
 2   costo_unitario      7 non-null      float64
 3   proveedor           7 non-null      object 
dtypes: float64(1), object(3)
memory usage: 352.0+ bytes


,nombre_producto,categoria_producto,costo_unitario,proveedor
0,Laptop-Gaming-16GB,Electrónica,280.68,"Fuller, Pena and Myers"
1,Phone-Pro-128GB,Electrónica,10.12,King Ltd
2,Tablet-Standard-64GB,Electrónica,25.21,Bowers LLC
3,Blender-XL-Red,Hogar,176.64,Long-Reid
4,Vacuum-Pro-Black,Hogar,16.60,"Rivera, Carr and Finley"
5,Sneakers-Urban-42,Moda,17.21,Greene-Smith
6,Jacket-Winter-M,Moda,189.31,Mcmillan-Rhodes


In [201]:
marketing.info()
marketing.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1620 entries, 0 to 1619
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   fecha       1620 non-null   object 
 1   pais        1620 non-null   object 
 2   id_campaña  1620 non-null   object 
 3   canal       1519 non-null   object 
 4   gasto       1620 non-null   float64
dtypes: float64(1), object(4)
memory usage: 63.4+ KB


,fecha,pais,id_campaña,canal,gasto
0,2025-01-01,Mexico,organic_Mexico,organic,2446.25
1,2025-01-01,Mexico,paid_search_Mexico,paid_search,2704.34
2,2025-01-01,Mexico,social_Mexico,social,2045.01
3,2025-01-01,Colombia,organic_Colombia,organic,2597.21
4,2025-01-01,Colombia,paid_search_Colombia,paid_search,1771.40
5,2025-01-01,Colombia,social_Colombia,social,833.38
6,2025-01-01,Argentina,organic_Argentina,organic,1365.62
7,2025-01-01,Argentina,paid_search_Argentina,paid_search,1282.65
8,2025-01-01,Argentina,social_Argentina,social,1534.69
9,2025-01-02,Mexico,organic_Mexico,organic,2537.38


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - La carga y exploración inicial de los datasets está organizada de forma clara y facilita comprender rápidamente la estructura de la información antes de iniciar el análisis.
</div>


---

### Revisión y calidad de datos

**🎯 Objetivo:** Detectar y corregir problemas en los datos que puedan afectar el análisis de revenue, costos y rentabilidad.

Se revisan los 3 datasets
- Validar y convertir fechas al formato correcto  
- Revisar variables numéricas (sin negativos o ceros inválidos)  
- Verificar consistencia de montos  
- Eliminar duplicados  
- Revisar variables categóricas 

---

In [202]:
catalog = catalog.rename(columns={'costo_unitario': 'precio_unitario'})

print()
catalog.info()
catalog.head(10)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   nombre_producto     7 non-null      object 
 1   categoria_producto  7 non-null      object 
 2   precio_unitario     7 non-null      float64
 3   proveedor           7 non-null      object 
dtypes: float64(1), object(3)
memory usage: 352.0+ bytes


,nombre_producto,categoria_producto,precio_unitario,proveedor
0,Laptop-Gaming-16GB,Electrónica,280.68,"Fuller, Pena and Myers"
1,Phone-Pro-128GB,Electrónica,10.12,King Ltd
2,Tablet-Standard-64GB,Electrónica,25.21,Bowers LLC
3,Blender-XL-Red,Hogar,176.64,Long-Reid
4,Vacuum-Pro-Black,Hogar,16.60,"Rivera, Carr and Finley"
5,Sneakers-Urban-42,Moda,17.21,Greene-Smith
6,Jacket-Winter-M,Moda,189.31,Mcmillan-Rhodes


In [203]:
orders['fecha_hora_pedido'] = pd.to_datetime(orders['fecha_hora_pedido'], errors='coerce')
orders['cantidad'] = orders['cantidad'].astype('Int64')
orders['pais'] = orders['pais'].str.title()
orders = orders.rename(columns={'fecha_hora_pedido': 'fecha_pedido'})
orders = orders.drop_duplicates()

precios_correctos = catalog[['nombre_producto', 'precio_unitario']].rename(
    columns={'precio_unitario': 'precio_correcto'}
)
orders = orders.merge(precios_correctos, on='nombre_producto', how='left')
orders['precio_unitario'] = orders['precio_correcto']
orders = orders.drop('precio_correcto', axis=1)

orders['cantidad'] = orders['cantidad'].fillna(1)
orders['monto_descuento'] = orders['monto_descuento'].fillna(0)

columnas_categoricas = ['pais', 'dispositivo', 'fuente_referencia', 'nombre_producto', 'categoria_producto']

for col in columnas_categoricas:
    orders[col] = orders[col].fillna('No especificado')

categorias_correctas = catalog[['nombre_producto', 'categoria_producto']]
orders = orders.merge(categorias_correctas, on='nombre_producto', how='left', suffixes=('', '_correcta'))

mask_no_especificado = orders['categoria_producto'] == 'No especificado'
orders.loc[mask_no_especificado, 'categoria_producto'] = orders.loc[mask_no_especificado, 'categoria_producto_correcta']

orders = orders.drop('categoria_producto_correcta', axis=1)

orders['monto_total'] = (
    orders['cantidad'] * orders['precio_unitario'] - 
    orders['monto_descuento']
)

orders = orders[~((orders['nombre_producto'] == 'No especificado') & (orders['precio_unitario'].isnull()))]

orders['categoria_producto'] = orders['categoria_producto'].replace({
    'Electronica': 'Electrónica'
})  

print()
orders.info()
orders.head(10)


<class 'pandas.core.frame.DataFrame'>
Int64Index: 24970 entries, 0 to 24999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   id_pedido           24970 non-null  object        
 1   id_usuario          24970 non-null  object        
 2   fecha_pedido        24970 non-null  datetime64[ns]
 3   pais                24970 non-null  object        
 4   dispositivo         24970 non-null  object        
 5   fuente_referencia   24970 non-null  object        
 6   nombre_producto     24970 non-null  object        
 7   categoria_producto  24970 non-null  object        
 8   cantidad            24970 non-null  Int64         
 9   precio_unitario     24970 non-null  float64       
 10  monto_descuento     24970 non-null  float64       
 11  monto_total         24970 non-null  Float64       
dtypes: Float64(1), Int64(1), datetime64[ns](1), float64(2), object(7)
memory usage: 2.5+ MB


,id_pedido,id_usuario,fecha_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
0,order_0,user_6993,2025-05-22,Argentina,desktop,organic,Jacket-Winter-M,Moda,2,189.31,0.0,378.62
1,order_1,user_1329,2025-06-15,Mexico,desktop,paid_search,Tablet-Standard-64GB,Electrónica,1,25.21,5.0,20.21
2,order_2,user_3194,2025-05-02,Argentina,desktop,social,Blender-XL-Red,Hogar,2,176.64,10.0,343.28
3,order_3,user_4510,2025-06-09,Colombia,mobile,social,Tablet-Standard-64GB,Electrónica,1,25.21,15.0,10.21
4,order_4,user_5044,2025-03-30,Argentina,desktop,paid_search,Blender-XL-Red,Hogar,1,176.64,0.0,176.64
5,order_5,user_2792,2025-04-22,Mexico,desktop,organic,Tablet-Standard-64GB,Electrónica,1,25.21,0.0,25.21
6,order_6,user_2802,2025-03-15,Colombia,mobile,organic,Blender-XL-Red,Hogar,2,176.64,0.0,353.28
7,order_7,user_1083,2025-05-01,Mexico,desktop,organic,Tablet-Standard-64GB,Electrónica,2,25.21,0.0,50.42
8,order_8,user_2352,2025-03-01,Argentina,desktop,paid_search,Laptop-Gaming-16GB,Electrónica,2,280.68,5.0,556.36
9,order_9,user_7001,2025-05-31,Mexico,mobile,organic,Sneakers-Urban-42,Moda,1,17.21,5.0,12.21


In [204]:
marketing['fecha'] = pd.to_datetime(marketing['fecha'], errors='coerce')
marketing['pais'] = marketing['pais'].str.title()
marketing = marketing.drop_duplicates()

def extraer_canal(id_campaña):
    if pd.isna(id_campaña):
        return 'No especificado'
    elif id_campaña.startswith('organic_'):
        return 'organic'
    elif id_campaña.startswith('paid_search_'):
        return 'paid_search'
    elif id_campaña.startswith('social_'):
        return 'social'
    else:
        return 'No especificado'

mask_canal_faltante = marketing['canal'].isna()
marketing.loc[mask_canal_faltante, 'canal'] = marketing.loc[mask_canal_faltante, 'id_campaña'].apply(extraer_canal)

marketing.info()
marketing.head(10)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1620 entries, 0 to 1619
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   fecha       1620 non-null   datetime64[ns]
 1   pais        1620 non-null   object        
 2   id_campaña  1620 non-null   object        
 3   canal       1620 non-null   object        
 4   gasto       1620 non-null   float64       
dtypes: datetime64[ns](1), float64(1), object(3)
memory usage: 75.9+ KB


,fecha,pais,id_campaña,canal,gasto
0,2025-01-01,Mexico,organic_Mexico,organic,2446.25
1,2025-01-01,Mexico,paid_search_Mexico,paid_search,2704.34
2,2025-01-01,Mexico,social_Mexico,social,2045.01
3,2025-01-01,Colombia,organic_Colombia,organic,2597.21
4,2025-01-01,Colombia,paid_search_Colombia,paid_search,1771.40
5,2025-01-01,Colombia,social_Colombia,social,833.38
6,2025-01-01,Argentina,organic_Argentina,organic,1365.62
7,2025-01-01,Argentina,paid_search_Argentina,paid_search,1282.65
8,2025-01-01,Argentina,social_Argentina,social,1534.69
9,2025-01-02,Mexico,organic_Mexico,organic,2537.38


---
**📦 Exportación**: Una vez finalizada la limpieza, se exportan los datasets para utilizarlos en la última etapa del proyecto.

In [205]:
# exportar datasets
orders.to_csv('orders_clean.csv', index=False)
catalog.to_csv('catalog_clean.csv', index=False)
marketing.to_csv('marketing_clean.csv', index=False)

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - La limpieza y validación de los datos está desarrollada con una lógica consistente y demuestra atención al detalle en la preparación del dataset para las siguientes etapas del análisis.
</div>


---

## 🔹 Paso 2: Analizar si el negocio es rentable

### 2.1 Cálculo de KPIs principales

**🎯 Objetivo:** Calcular los indicadores clave del negocio para evaluar ingresos, costos y rentabilidad.

Se usan los 3 datasets (`orders`, `catalog`, `marketing_spend`):

**📊 Parte 1: Rentabilidad del negocio**
- ¿Cuál es el ingreso total (revenue)? 
- ¿Cuál es el costo total? 
- ¿Cuánto se ha invertido en marketing? 
- ¿El negocio es rentable? (calcular profit)  

---

**📈 Parte 2: Comportamiento de ventas**
- ¿Cuál es el ticket promedio por orden? 
- ¿Cuál es la cantidad promedio de productos por orden? 
- ¿Cuál es el producto más vendido?
- ¿Cuánto se ha gastado en marketing por canal? 

In [206]:
print("RENTABILIDAD DEL NEGOCIO")

print("\nIngreso Total (Revenue)")
revenue = orders['monto_total'].sum()
print(f"${revenue:,.2f}")

print("\nCostos Totales")
costos = orders['cantidad'] * orders['precio_unitario']
costo_total = costos.sum()
print(f"${costo_total:,.2f}")

print("\nInversión de marketing total")
inversion_marketing = marketing['gasto'].sum()
print(f"${inversion_marketing:,.2f}")

print("\n¿El negocio es rentable?")
profit = revenue - costo_total - inversion_marketing
print(f"Profit = Revenue - Costos - Marketing")
print(f"Profit = ${profit:,.2f}")

if profit > 0:
    print("✅ El negocio ES rentable")
else:
    print("❌ El negocio NO es rentable")

RENTABILIDAD DEL NEGOCIO

Ingreso Total (Revenue)
$43,015,932.15

Costos Totales
$43,128,147.15

Inversión de marketing total
$2,871,843.53

¿El negocio es rentable?
Profit = Revenue - Costos - Marketing
Profit = $-2,984,058.53
❌ El negocio NO es rentable


In [207]:
print("COMPORTAMIENTO DE VENTAS")

print("\nTicket Promedio")
ticket_mean = orders['monto_total'].mean()
print(f"${ticket_mean:,.2f}")

print("\nProductos promedio por orden")
order_mean = orders['cantidad'].mean()
print(f"{order_mean:,.2f}")

print("\nProducto más Vendido")
product_mas = orders.groupby('nombre_producto')['cantidad'].sum().sort_values(ascending=False)
print(product_mas)

print("\nGastos totales por canal de Marketing")
gasto_canal = marketing.groupby('canal')['gasto'].sum().sort_values(ascending=False)
for canal, gasto in gasto_canal.items():
    print(f"{canal}: ${gasto:,.2f}")

COMPORTAMIENTO DE VENTAS

Ticket Promedio
$1,722.70

Productos promedio por orden
7.11

Producto más Vendido
nombre_producto
Laptop-Gaming-16GB      144202
Vacuum-Pro-Black          6290
Blender-XL-Red            6282
Jacket-Winter-M           6266
Sneakers-Urban-42         6191
Tablet-Standard-64GB      4158
Phone-Pro-128GB           4138
Name: cantidad, dtype: Int64

Gastos totales por canal de Marketing
social: $976,818.37
organic: $972,650.96
paid_search: $922,374.20


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - El cálculo de los KPIs principales está presentado de forma clara y permite interpretar fácilmente la situación financiera y el comportamiento general de las ventas del negocio.
</div>


---

## 🔹 Paso 3: Entender dónde se pierden los usuarios (funnel de conversión)

**🎯 Objetivo:** Analizar el comportamiento de los usuarios para identificar en qué etapa del proceso se pierden.


⚙️**Conexión a la base de datos**:  
Se ejecuta la línea de configuración para conectar con la base de datos y aplicar consultas SQL en la tabla **events**.

---

**📊 Parte 1: Construcción del funnel**
- ¿Cuántos usuarios llegan a cada etapa del funnel?  
- Se calcula el número de usuarios únicos por `nombre_evento`  
- Se ordenan los eventos según el flujo del usuario  

---

**📉 Parte 2: Análisis de conversión**
- Se calcula la tasa de conversión entre cada paso del funnel  
- Se identifica en qué etapa se pierde la mayor cantidad de usuarios  
- ¿Cuál es la tasa de conversión final?
---

In [208]:
import pandas as pd
from sqlalchemy import create_engine

# ======================
# Conexión (NO modificar)
# ======================
db_config = {
    'user': 'practicum_student',
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7',
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
    'port': 5432,
    'db': 'data-analyst-production-db-en'
}

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [209]:
# Explorar tabla events
# =========================
query_events = '''
SELECT *
FROM events;
'''
events = pd.read_sql(query_events, con=engine)
events.head(10)

,id_usuario,id_sesion,nombre_evento,timestamp_evento,pais,dispositivo,fuente_referencia,categoria_producto
0,user_6772,6a97f2af-32ae-4186-8c92-04025be1a27b,first_visit,2025-05-17,Colombia,desktop,organic,Moda
1,user_5883,369b767c-1c33-4b2f-a652-c7c0ef92cfc9,add_to_cart,2025-02-23,Mexico,mobile,social,Hogar
2,user_5946,60039041-e78b-474c-87b3-c0b7e9c30708,add_payment_info,2025-05-15,Colombia,desktop,social,Electronica
3,user_827,18252a64-f389-4ef7-9e58-dadad4a3491e,purchase,2025-03-31,Mexico,mobile,social,Moda
4,user_2361,221b364e-cdc5-4668-b698-18d5ba849a67,first_visit,2025-01-22,Argentina,desktop,paid_search,Electronica
5,user_4402,0892643c-f80d-4eb4-9ca8-3638b16c6ed4,begin_checkout,2025-04-21,Argentina,desktop,paid_search,Hogar
6,user_767,0005a13c-3c1f-496b-a26a-f3c282f19e13,begin_checkout,2025-03-01,Mexico,mobile,social,Electronica
7,user_2694,b07336e6-1207-4ee1-82e3-1fa7fbd194e3,add_payment_info,2025-03-22,Colombia,desktop,organic,Hogar
8,user_7209,f42bd8cd-aff2-4b7e-94e2-d4d885d23afc,add_to_cart,2025-05-07,Mexico,desktop,organic,Moda
9,user_1788,abf1af77-e96b-4490-9efb-3fa5d8271ad4,add_to_cart,2025-05-13,Argentina,desktop,organic,Moda


In [210]:
events.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120000 entries, 0 to 119999
Data columns (total 8 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   id_usuario          120000 non-null  object
 1   id_sesion           120000 non-null  object
 2   nombre_evento       120000 non-null  object
 3   timestamp_evento    120000 non-null  object
 4   pais                120000 non-null  object
 5   dispositivo         120000 non-null  object
 6   fuente_referencia   120000 non-null  object
 7   categoria_producto  120000 non-null  object
dtypes: object(8)
memory usage: 7.3+ MB


In [211]:
# PARTE 1: Totales del funnel
# ======================

query_totals = '''
WITH cte_first AS (
  SELECT DISTINCT id_usuario
  FROM events
  WHERE nombre_evento = 'first_visit'
),
cte_item AS (
  SELECT DISTINCT id_usuario
  FROM events
  WHERE nombre_evento = 'select_item'
),
cte_cart AS (
  SELECT DISTINCT id_usuario
  FROM events
  WHERE nombre_evento = 'add_to_cart'
),
cte_check AS (
  SELECT DISTINCT id_usuario
  FROM events
  WHERE nombre_evento = 'begin_checkout'
),
cte_payment_info AS (
  SELECT DISTINCT id_usuario
  FROM events
  WHERE nombre_evento = 'add_payment_info'
),
cte_purch AS (
  SELECT DISTINCT id_usuario
  FROM events
  WHERE nombre_evento = 'purchase'
)
SELECT
  (SELECT COUNT(*) FROM cte_first) AS first_visit_users,
  (SELECT COUNT(*) FROM cte_item) AS select_item_users,
  (SELECT COUNT(*) FROM cte_cart) AS cart_users,
  (SELECT COUNT(*) FROM cte_check) AS checkout_users,
  (SELECT COUNT(*) FROM cte_payment_info) AS payment_info_users,
  (SELECT COUNT(*) FROM cte_purch) AS purchase_users;
'''

totals = pd.read_sql(query_totals, con=engine)
totals

,first_visit_users,select_item_users,cart_users,checkout_users,payment_info_users,purchase_users
0,7796,7582,7634,7208,6250,6240


In [212]:
# PARTE 2: Conversiones
# ======================

query_conversion = '''

WITH cte_first AS (
  SELECT DISTINCT id_usuario
  FROM events
  WHERE nombre_evento = 'first_visit'
),
cte_item AS (
  SELECT DISTINCT id_usuario
  FROM events
  WHERE nombre_evento = 'select_item'
),
cte_cart AS (
  SELECT DISTINCT id_usuario
  FROM events
  WHERE nombre_evento = 'add_to_cart'
),
cte_check AS (
  SELECT DISTINCT id_usuario
  FROM events
  WHERE nombre_evento = 'begin_checkout'
),
cte_payment_info AS (
  SELECT DISTINCT id_usuario
  FROM events
  WHERE nombre_evento = 'add_payment_info'
),
cte_purch AS (
  SELECT DISTINCT id_usuario
  FROM events
  WHERE nombre_evento = 'purchase'
)
SELECT
  (SELECT COUNT(*) FROM cte_first) AS first_visit_users,
  (SELECT COUNT(*) FROM cte_item) AS select_item_users,
  (SELECT COUNT(*) FROM cte_cart) AS cart_users,
  (SELECT COUNT(*) FROM cte_check) AS checkout_users,
  (SELECT COUNT(*) FROM cte_payment_info) AS payment_info_users,
  (SELECT COUNT(*) FROM cte_purch) AS purchase_users,
  
((SELECT COUNT(*) FROM cte_first) - (SELECT COUNT(*) FROM cte_item)) * 100 
    / NULLIF((SELECT COUNT(*) FROM cte_first), 0) AS dropoff_first_visit_pct,
    
    ((SELECT COUNT(*) FROM cte_item) - (SELECT COUNT(*) FROM cte_cart)) * 100 
    / NULLIF((SELECT COUNT(*) FROM cte_item), 0) AS dropoff_item_pct,
       
    ((SELECT COUNT(*) FROM cte_cart) - (SELECT COUNT(*) FROM cte_check)) * 100 
    / NULLIF((SELECT COUNT(*) FROM cte_cart), 0) AS dropoff_cart_pct,
    
    ((SELECT COUNT(*) FROM cte_check) - (SELECT COUNT(*) FROM cte_payment_info)) * 100 
    / NULLIF((SELECT COUNT(*) FROM cte_check), 0) AS dropoff_check_pct,
    
    ((SELECT COUNT(*) FROM cte_payment_info) - (SELECT COUNT(*) FROM cte_purch)) * 100 
    / NULLIF((SELECT COUNT(*) FROM cte_payment_info), 0) AS dropoff_payment_pct;

'''

conversion = pd.read_sql(query_conversion, con=engine)
conversion

,first_visit_users,select_item_users,cart_users,checkout_users,payment_info_users,purchase_users,dropoff_first_visit_pct,dropoff_item_pct,dropoff_cart_pct,dropoff_check_pct,dropoff_payment_pct
0,7796,7582,7634,7208,6250,6240,2,0,5,13,0


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - La construcción del funnel y el análisis de conversión muestran una estructura clara y permiten identificar con facilidad las etapas críticas del recorrido de los usuarios.
</div>


---

## 🔹 Paso 4: Evaluar si los usuarios regresan (retención por cohortes)

**🎯 Objetivo:** Analizar la retención de usuarios para entender si regresan después de registrarse.

**Tablas**

- `users` 
- `user_activity` 

---
1. Se identifica la cohorte de cada usuario según el **mes de registro**.


2. Se calcula la retención semanal: cuántos usuarios **se mantienen activos** en cada semana desde su registro.
   - `retenido_w1`: usuarios activos en la semana 1  
   - `retenido_w2`: usuarios activos en la semana 2  
   - `retenido_w3`: usuarios activos en la semana 3  


3. Se calcula el porcentaje de retención para cada semana, dividiendo los usuarios retenidos entre los clientes iniciales de la cohorte:  
   - `semana_1`: porcentaje de usuarios retenidos en la semana 1  
   - `semana_2`: porcentaje de usuarios retenidos en la semana 2  
   - `semana_3`: porcentaje de usuarios retenidos en la semana 3  

Se revisa que la columna de fecha esté en formato correcto (`DATE`).  
Se realiza la conversión usando: `CAST(fecha_registro AS DATE)`

In [213]:
# Explorar tabla users
# =========================
query_users = '''
SELECT 
    id_usuario,
    CAST(fecha_registro AS DATE) as fecha_registro,
    país,
    dispositivo,
    tipo_plan
FROM users;
'''
users = pd.read_sql(query_users, con=engine)
users.info()
users.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   id_usuario      8000 non-null   object
 1   fecha_registro  8000 non-null   object
 2   país            8000 non-null   object
 3   dispositivo     8000 non-null   object
 4   tipo_plan       8000 non-null   object
dtypes: object(5)
memory usage: 312.6+ KB


,id_usuario,fecha_registro,país,dispositivo,tipo_plan
0,user_0,2025-01-29,Mexico,mobile,free
1,user_1,2025-01-07,Mexico,mobile,free
2,user_2,2025-03-12,Argentina,mobile,free
3,user_3,2025-03-04,Mexico,desktop,free
4,user_4,2025-02-27,Argentina,desktop,free


In [214]:
# Explorar tabla user_activity
# =========================
query_user_activity = '''
SELECT 
    id_usuario,
    CAST(fecha_actividad AS DATE) as fecha_actividad,
    dias_despues_registro,
    activo
FROM user_activity;
'''
user_activity = pd.read_sql(query_user_activity, con=engine)
user_activity.info()
user_activity.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32000 entries, 0 to 31999
Data columns (total 4 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   id_usuario             32000 non-null  object
 1   fecha_actividad        32000 non-null  object
 2   dias_despues_registro  32000 non-null  int64 
 3   activo                 32000 non-null  int64 
dtypes: int64(2), object(2)
memory usage: 1000.1+ KB


,id_usuario,fecha_actividad,dias_despues_registro,activo
0,user_0,2025-02-05,7,0
1,user_0,2025-02-12,14,1
2,user_0,2025-02-19,21,1
3,user_0,2025-02-26,28,0
4,user_1,2025-01-14,7,0


In [215]:

# Retención por cohortes
# ======================

query_cohort_retention_final = '''
WITH cohortes AS (
    SELECT
        u.id_usuario,
        DATE_TRUNC('week', CAST(u.fecha_registro AS DATE)) AS cohorte_semana,
        ua.dias_despues_registro,
        ua.activo
    FROM users u
    JOIN user_activity ua ON u.id_usuario = ua.id_usuario
),

usuarios_por_cohorte AS (
    SELECT 
        cohorte_semana,
        COUNT(DISTINCT id_usuario) AS clientes_iniciales
    FROM cohortes
    GROUP BY cohorte_semana
),

retencion AS (
    SELECT
        cohorte_semana, 
        COUNT(CASE WHEN dias_despues_registro BETWEEN 1 AND 7 AND activo = 1 THEN id_usuario END) AS retained_s1,
        COUNT(CASE WHEN dias_despues_registro BETWEEN 8 AND 14 AND activo = 1 THEN id_usuario END) AS retained_s2,
        COUNT(CASE WHEN dias_despues_registro BETWEEN 15 AND 21 AND activo = 1 THEN id_usuario END) AS retained_s3
    FROM cohortes
    GROUP BY cohorte_semana
)

SELECT
    TO_CHAR(r.cohorte_semana, 'YYYY-MM-DD') AS cohorte,
    u.clientes_iniciales,
    ROUND(r.retained_s1::numeric / u.clientes_iniciales, 2) AS semana_1,
    ROUND(r.retained_s2::numeric / u.clientes_iniciales, 2) AS semana_2,
    ROUND(r.retained_s3::numeric / u.clientes_iniciales, 2) AS semana_3
FROM retencion r
JOIN usuarios_por_cohorte u ON r.cohorte_semana = u.cohorte_semana
ORDER BY r.cohorte_semana;
'''

# Ejecutar la consulta
cohorte_final = pd.read_sql(query_cohort_retention_final, con=engine)
cohorte_final

,cohorte,clientes_iniciales,semana_1,semana_2,semana_3
0,2024-12-30,236,0.42,0.39,0.40
1,2025-01-06,351,0.41,0.45,0.42
2,2025-01-13,362,0.44,0.38,0.42
3,2025-01-20,394,0.44,0.40,0.40
4,2025-01-27,373,0.42,0.47,0.40
5,2025-02-03,405,0.40,0.44,0.45
6,2025-02-10,364,0.42,0.40,0.45
7,2025-02-17,338,0.47,0.40,0.40
8,2025-02-24,353,0.42,0.42,0.43
9,2025-03-03,363,0.40,0.44,0.40


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - El análisis de cohortes está bien estructurado y permite interpretar de manera clara el comportamiento de retención de los usuarios a lo largo del tiempo.
</div>


---

## 🔹 Paso 5: Validar si los cambios generan impacto (test estadístico)

🎯 **Objetivo:** Evaluar si la modificación en la UI del checkout impacta la **tasa de conversión de compra**.

---

1. **Analizar el dataset** `experiment_checkout_ui.csv` para identificar la métrica principal **conversion**.
   - La métrica **conversion** es 1 si el usuario completó la compra, 0 si no.    
2. **Plantear la hipótesis estadística**     
3. **Aplicar el test estadístico adecuado** 
4. **Interpretar el resultado**  

---
Hipótesis estadística
   - **H₀ (Hipótesis nula):** La `UI Experimental` no presenta un impacto en las conversiones finales en comparación con la`UI Actual`.
   - **H₁ (Hipótesis alternativa):** La `UI Experimental` presenta un impacto diferente (mayor o menor) en las conversiones finales comparada con la `UI Actual`.
   
**Test estadístico:** 
Estadístico t: 0.81322, Valor P: 0.41611

**Nivel de significancia alpha:** 0.05

In [216]:
# tu código aquí
import pandas as pd
from scipy.stats import ttest_ind
chkout = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/experiment_checkout_ui.csv')
chkout.info()
chkout.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id_usuario       10000 non-null  object 
 1   variante         10000 non-null  object 
 2   convirtio        10000 non-null  int64  
 3   dispositivo      10000 non-null  object 
 4   pais             10000 non-null  object 
 5   duracion_sesion  10000 non-null  float64
 6   timestamp        10000 non-null  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 547.0+ KB


,id_usuario,variante,convirtio,dispositivo,pais,duracion_sesion,timestamp
0,exp_user_0,tratamiento,0,mobile,Argentina,114.41,2025-03-28
1,exp_user_1,tratamiento,0,desktop,Mexico,170.03,2025-01-15
2,exp_user_2,control,1,mobile,Colombia,140.21,2025-03-18
3,exp_user_3,tratamiento,0,mobile,Colombia,151.45,2025-06-03
4,exp_user_4,tratamiento,0,desktop,Mexico,299.96,2025-01-12


In [217]:
tratamiento1 = chkout[(chkout['variante'] == 'tratamiento') & (chkout['convirtio'] == 1)]
tratamiento0 = chkout[(chkout['variante'] == 'tratamiento') & (chkout['convirtio'] == 0)]
control1 = chkout[(chkout['variante'] == 'control') & (chkout['convirtio'] == 1)]
control0 = chkout[(chkout['variante'] == 'control') & (chkout['convirtio'] == 0)]

print(f"Tratamiento - Convirtieron: {len(tratamiento1)}")
print(f"Tratamiento - No convirtieron: {len(tratamiento0)}")
print(f"Control - Convirtieron: {len(control1)}")
print(f"Control - No convirtieron: {len(control0)}")

tasa_tratamiento = len(tratamiento1) / (len(tratamiento1) + len(tratamiento0))
tasa_control = len(control1) / (len(control1) + len(control0))


print(f"\nTasas de conversión:")
print(f"Tratamiento: {tasa_tratamiento:.3f} ({tasa_tratamiento*100:.1f}%)")
print(f"Control: {tasa_control:.3f} ({tasa_control*100:.1f}%)")
print(f"Diferencia: {tasa_tratamiento - tasa_control:.3f}")

Tratamiento - Convirtieron: 820
Tratamiento - No convirtieron: 4215
Control - Convirtieron: 779
Control - No convirtieron: 4186

Tasas de conversión:
Tratamiento: 0.163 (16.3%)
Control: 0.157 (15.7%)
Diferencia: 0.006


In [218]:
tratamiento = chkout[chkout['variante'] == 'tratamiento']['convirtio']
control = chkout[chkout['variante'] == 'control']['convirtio']

ttest_ind(tratamiento, control)

t_stat, p_value = ttest_ind(tratamiento, control)

print(f"Estadístico t: {t_stat:.5f}")
print(f"Valor P: {p_value:.5f}")

if p_value < 0.05:
    print("✅ Diferencia estadísticamente significativa")
else:
    print("❌ No hay diferencia estadísticamente significativa")


Estadístico t: 0.81322
Valor P: 0.41611
❌ No hay diferencia estadísticamente significativa


**INTERPRETACIÓN**

* H0: No hay diferencia real entre la nueva UI y la UI actual".
* Resultado: No hay evidencia estadística suficiente para decir que la nueva UI es mejor.
* Interpretación: La diferencia observada (16.3% vs 15.7%) podría ser simplemente casualidad. 

**RECOMENDACIÓN**

Con un p-value de 0.416, NO se recomienda implementar la nueva UI debido a que no se pueden asegurar mejoras constantes.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - La validación estadística está bien desarrollada y la interpretación de los resultados permite conectar correctamente el análisis técnico con una decisión de negocio clara.
</div>


---

## 🔹 Paso 6: Comunicar los resultados (Dashboard en BI)

🎯 **Objetivo**:  
Crear un dashboard que muestre de manera clara y visual los resultados del análisis de ventas, costos, marketing y conversión. 

Se usarán los CSVs limpios del Paso 1:

- `orders_clean.csv`  
- `catalog_clean.csv`  
- `marketing_clean.csv`

---

1️⃣ Preparación de los datos
1. Cargar los CSVs en Power BI o Tableau.
2. Revisar relaciones:
   - `orders.nombre_producto` → `catalog.nombre_producto`
   - `orders.fecha_pedido` → tabla de fechas (crear calendario para análisis temporal)
   - `orders.fecha_pedido` → `dim_fecha.date`
3. Crear columnas calculadas necesarias
4. Crear tabla de fechas para poder calcular comparaciones YTD, YoY o períodos anteriores (`Previous Year`, `Previous Month`).

---

2️⃣ Dashboard 1: Overview Ejecutivo
**KPIs principales a mostrar:**
- Revenue total
- Profit total
- Gasto total en marketing
- Ticket promedio
- Cantidad promedio de productos por orden

**Visualizaciones sugeridas:**
- Tarjetas KPI para revenue, profit y gasto marketing
- Gráfico de líneas: evolución mensual de revenue o profit
- Gráfico de líneas YTD
- Gráfico de barras: revenue y profit por producto o categoría

---

 3️⃣ Dashboard 2: Detalle / Drill-through  
**Objetivo:** Permitir explorar los datos desde el KPI general hasta cada orden o producto.

**Visualizaciones sugeridas:**
- Tabla detallada de órdenes con:
  - producto, cantidad, revenue, cost, profit
  - color condicional (profit negativo en rojo, positivo en verde)
- Gráfico de barras por producto con medida `cantidad vendida`
- Drill-through: seleccionar un producto y ver todos los pedidos relacionados
- Filtros por fecha, categoría de producto, etc

---

## 🚀 Entrega Final

Comparte el acceso a tu Dashboard para revisión.   
Puedes entregar el Dashboard utilizando **Power BI o Tableau**.

Incluye **uno de los siguientes**:

- 🔗 Link público del dashboard publicado en **Power BI Service o Tableau Public / Tableau Cloud**
- 🔗 Link de **Google Drive o OneDrive** con el archivo del proyecto (`.pbix`) y los 3 csvs limpios.


### 📎 Enlace del Dashboard

https://public.tableau.com/app/profile/rodrigo.salgado.galindo/viz/Sprint12-ProyectoFinalRappiPlus/Overview?publish=yes

# Comentario General del Revisor

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

El proyecto presenta un flujo analítico sólido y bien conectado con las necesidades del negocio planteadas. La estructura del notebook permite seguir cada etapa del análisis de manera ordenada, desde la validación de datos hasta la interpretación estadística y visual de resultados.

#### Evaluación del código:

* **Preparación y limpieza de datos:** Se observa una limpieza consistente, manejo adecuado de tipos de datos, normalización de variables categóricas y validación lógica de métricas financieras.
* **Cálculo de KPIs:** Los indicadores principales fueron desarrollados correctamente y ayudan a evaluar de forma clara la rentabilidad del negocio.
* **Consultas SQL y funnel:** La construcción del funnel y las consultas mediante CTEs muestran una buena organización lógica y permiten interpretar correctamente las pérdidas de usuarios en cada etapa.
* **Análisis de cohortes:** La retención semanal está correctamente estructurada y facilita analizar el comportamiento de permanencia de los usuarios.
* **Test estadístico:** La formulación de hipótesis, aplicación del test t e interpretación del p-value están alineadas con el objetivo del experimento y con la toma de decisiones de negocio.

#### Evaluación del dashboard:

* **Organización visual:** El dashboard resume de forma clara los principales indicadores financieros y facilita una lectura rápida de revenue, costos y profit.
* **Visualización de tendencias:** La evolución mensual de ingresos permite identificar cambios relevantes en el desempeño del negocio durante el período analizado.
* **Comparación por categorías:** La separación entre revenue y profit por categoría aporta contexto útil para identificar áreas con mejor o peor desempeño.
* **Uso de filtros:** La incorporación de filtros mejora la exploración interactiva de la información y aporta flexibilidad al análisis.

El proyecto cumple con los objetivos principales  y logra conectar correctamente el análisis técnico con decisiones de negocio basadas en datos.

</div>
